Update CI workflows

rokm opened the pull request:
Make python 3.13 first-class citizen of our `tests` CI workflow; i.e., upgrade from `3.13-dev` to `3.13` to run with all libraries. Well, except `gevent` and `PySide6`, which do not support python 3.13 yet.

`macos-12` runner is being phased out, so switch to a newer one. `macos-14` is arm64 and is *significantly* faster than x86_64-based `macos-12` or `macos-13`. 

However, it might be a good idea to retain at least some tests with x86_64 - in that case, we will need to also add at least some `macos-13` runners. Perhaps run python <= 3.10 with macos-13 and python > 3.10 with macos-14, or something along those lines, to avoid running all versions on both (which is slow again - see [here](https://github.com/rokm/pyinstaller/actions/runs/11278858588)). As a side note, trying to run x86_64 python through rosetta2 on macos-14 has no speed benefit over macos-13 (see [here](https://github.com/rokm/pyinstaller/actions/runs/11274181176)), and it brings additional problems (the test failures are presumably due to lack of Homebrew-provided x86_64 OpenSSL library).

rokm committed:
tests: PySide6 does not support python 3.13 yet

rokm committed:
tests: gevent does not support python 3.13 yet

While building the wheels seems to succeed, the resulting build
is defunct due to lack of proper support for python 3.13.

rokm committed:
tests: PySide2 wheels are not available for arm64 on macOS

rokm committed:
tests: remove keyring pin for old python + macOS combination

rokm committed:
tests: apple events: fix path to Carbon.framework

Instead of `/System/Library/Carbon.framework/Carbon`, the path
to the `Carbon.framework` shared library should be
`/System/Library/Frameworks/Carbon.framework/Carbon`.

With python 3.8 - 3.12, `ctypes.CDLL` was able to resolve it
nevertheless (and would also resolve just the framework/library
name without path, e.g., `Carbon.framework/Carbon`); but with
python 3.13, this is no longer seems to be the case.

rokm committed:
tests: apple events: use UUID for unique URL scheme and file extension

Use UUID for unique URL scheme and file extension, instead of
`int(time.time())`; with the latter, we get collisions when
running just Apple events test on multiple workers and different
variants of the test end up being started in the same second.

rokm committed:
ci: run full test workflow with python 3.13

Use python 3.13 instead of 3.13-dev to run the full workflow with
it.

rokm committed:
ci: run macOS workflow on macos-14 runner

bwoodsend commented:
Some Python versions on `x86_64` and some on `arm64` sounds sensible to me. I'd personally shuffle them a bit so it's not all the old ones on one and new ones on the other.

rokm committed:
ci: split macOS test workflows between macos-13 and macos-14 runners

Split macOS workflows between macos-13 (x86_64) and macos-14 (arm64)
runners to cover both architectures.

bwoodsend commented:
Looks like I'll need to go update that silly list of job names in the branch protection rules...

rokm commented:
> Looks like I'll need to go update that silly list of job names in the branch protection rules...

I've updated the list.
